IGV report of the mapping results of assembled contigs. Source reads of trait-associated k-mers were assembled using `Spades <https://github.com/ablab/spades>`_. Resulted contigs were mapped to the reference genome using `minimap2 <https://github.com/lh3/minimap2>`_. Mapping results were filtered based minimum mapping quality score of {{ snakemake.config["params"]["filter_alignment"]["min_map_score"] }}. Alignment BAM files were converted to BED format using `bedtools <https://bedtools.readthedocs.io/en/latest/>`_. IGV HTML report was generated using `igv-reports <https://github.com/igvteam/igv-reports>`_. IGV report consists of alignment BAM and BED files, the reference genome FASTA and the reference gene annotations in GTF/GFF3.