kmersGWAS results for the phenotype summary table. k-mers-based GWAS was performed by following `the method of Voichek et al. (2020) <https://github.com/voichek/kmersGWAS/blob/master/manual.pdf>`_. k-mers length was set to k= {{ snakemake.config["params"]["kmc"]["kmer_len"] }}. The first column represents the associated k-mers that pass the p-value threshold (5% or 10%), the second column shows the phenotype, the third column shows the allele frequency of a k-mer in that row, and the next three columns represent the p-value, -log10 transformed p-value, and the permutation based p-value threshold (5% or 10%) respectively.